app/platform/metrics.go

```diff
@@ -0,0 +1,162 @@
+	"github.com/mattermost/mattermost-server/v6/shared/mlog"
+	"github.com/pkg/errors"
+)
+
+const TimeToWaitForConnectionsToCloseOnServerShutdown = time.Second
+
+type platformMetrics struct {
+	server *http.Server
+	router *mux.Router
+	lock   sync.Mutex
+
+	metricsImpl einterfaces.MetricsInterface
+
+	cfgFn func() *model.Config
+}
+
+func newPlatformMetrics(metricsImpl einterfaces.MetricsInterface, cfgFn func() *model.Config) *platformMetrics {
+	if !*cfgFn().MetricsSettings.Enable {
+		return nil
+	}
+
+	pm := &platformMetrics{
+		cfgFn: cfgFn,
+	}
+
+	pm.stopMetricsServer()
+
+	if err := pm.initMetricsRouter(); err != nil {
+		mlog.Error("Error initiating metrics router.", mlog.Err(err))
+	}
+
+	if metricsImpl != nil {
+		metricsImpl.Register()
+	}
+
+	pm.startMetricsServer()
+
+	return pm
+}
+
+func (pm *platformMetrics) stopMetricsServer() {
+	pm.lock.Lock()
+	defer pm.lock.Unlock()
+
+	if pm.server != nil {
+		ctx, cancel := context.WithTimeout(context.Background(), TimeToWaitForConnectionsToCloseOnServerShutdown)
+		defer cancel()
+
+		pm.server.Shutdown(ctx)
+		mlog.Info("Metrics and profiling server is stopping")
+	}
+}
+
+func (pm *platformMetrics) startMetricsServer() {
+	var notify chan struct{}
+	pm.lock.Lock()
+	defer func() {
+		if notify != nil {
+			<-notify
+		}
+		pm.lock.Unlock()
+	}()
+
+	l, err := net.Listen("tcp", *pm.cfgFn().MetricsSettings.ListenAddress)
+	if err != nil {
+		mlog.Error(err.Error())
+		return
+	}
+
+	notify = make(chan struct{})
+	pm.server = &http.Server{
+		Handler:      handlers.RecoveryHandler(handlers.PrintRecoveryStack(true))(pm.router),
+		ReadTimeout:  time.Duration(*pm.cfgFn().ServiceSettings.ReadTimeout) * time.Second,
+		WriteTimeout: time.Duration(*pm.cfgFn().ServiceSettings.WriteTimeout) * time.Second,
+	}
+
+	go func() {
+		close(notify)
+		if err := pm.server.Serve(l); err != nil && err != http.ErrServerClosed {
+			mlog.Critical(err.Error())
+		}
+	}()
+
+	mlog.Info("Metrics and profiling server is started", mlog.String("address", l.Addr().String()))
+}
+
+func (pm *platformMetrics) initMetricsRouter() error {
+	pm.router = mux.NewRouter()
+	runtime.SetBlockProfileRate(*pm.cfgFn().MetricsSettings.BlockProfileRate)
+
+	metricsPage := `
+			<html>
+				<body>{{if .}}
+					<div><a href="/metrics">Metrics</a></div>{{end}}
+					<div><a href="/debug/pprof/">Profiling Root</a></div>
+					<div><a href="/debug/pprof/cmdline">Profiling Command Line</a></div>
+					<div><a href="/debug/pprof/symbol">Profiling Symbols</a></div>
+					<div><a href="/debug/pprof/goroutine">Profiling Goroutines</a></div>
+					<div><a href="/debug/pprof/heap">Profiling Heap</a></div>
+					<div><a href="/debug/pprof/threadcreate">Profiling Threads</a></div>
+					<div><a href="/debug/pprof/block">Profiling Blocking</a></div>
+					<div><a href="/debug/pprof/trace">Profiling Execution Trace</a></div>
+					<div><a href="/debug/pprof/profile">Profiling CPU</a></div>
+				</body>
+			</html>
```

Comment:
Would it make sense to embed this from an external file, rather than inlining a whole HTML document?

Reply:
Yeah, maybe we can do that in a separate PR, maybe we can also add some styling :p